pandas/tools/merge.py

```diff
@@ -1030,36 +1006,78 @@ def _get_merge_keys(self):
     def _get_join_indexers(self):
         """ return the join indexers """
 
+        # values to compare
+        left_values = self.left_join_keys[-1]
+        right_values = self.right_join_keys[-1]
+        tolerance = self.tolerance
+
         # we required sortedness in the join keys
         msg = " keys must be sorted"
-        for lk in self.left_join_keys:
-            if not Index(lk).is_monotonic:
-                raise ValueError('left' + msg)
-        for rk in self.right_join_keys:
-            if not Index(rk).is_monotonic:
-                raise ValueError('right' + msg)
-
-        kwargs = {}
-
-        # tolerance
-        t = self.tolerance
-        if t is not None:
-            lt = self.left_join_keys[self.left_on.index(self._asof_key)]
-            rt = self.right_join_keys[self.right_on.index(self._asof_key)]
-            if needs_i8_conversion(lt):
-                lt = lt.view('i8')
-                t = t.value
-                rt = rt.view('i8')
-            kwargs['left_values'] = lt
-            kwargs['right_values'] = rt
-            kwargs['tolerance'] = t
+        if not Index(left_values).is_monotonic:
+            raise ValueError('left' + msg)
+        if not Index(right_values).is_monotonic:
+            raise ValueError('right' + msg)
+
+        # initial type conversion as needed
+        if needs_i8_conversion(left_values):
+            left_values = left_values.view('i8')
+            right_values = right_values.view('i8')
+            if tolerance is not None:
+                tolerance = tolerance.value
+
+        # a "by" parameter requires special handling
+        if self.by is not None:
+            left_by_values = self.left_join_keys[0]
+            right_by_values = self.right_join_keys[0]
+
+            # choose appropriate function by type
+            on_dtype = left_values.dtype
+            by_dtype = left_by_values.dtype
+            if is_integer_dtype(by_dtype):
```

Comment:
see core/algorithms.py, basically put tuple of the args in a dict then get it

e.g.

```
_asof_funcs = { ('int64', 'double') : .....,
                         ('int64', 'int64'): ....}
```